Variadic tuple types

ahejlsberg opened the pull request:
This PR implements _variadic tuple types_, i.e. the ability for tuple types to have spreads of generic types that can be replaced with actual elements through type instantiation. The PR effectively implements the features discussed in #5453.

Some examples of new capabilities provided in this PR:

```ts
// Variadic tuple elements

type Foo<T extends unknown[]> = [string, ...T, number];

type T1 = Foo<[boolean]>;  // [string, boolean, number]
type T2 = Foo<[number, number]>;  // [string, number, number, number]
type T3 = Foo<[]>;  // [string, number]

// Strongly typed tuple concatenation

function concat<T extends unknown[], U extends unknown[]>(t: [...T], u: [...U]): [...T, ...U] {
    return [...t, ...u];
}

const ns = [0, 1, 2, 3];  // number[]

const t1 = concat([1, 2], ['hello']);  // [number, number, string]
const t2 = concat([true], t1);  // [boolean, number, number, string]
const t3 = concat([true], ns);  // [boolean, ...number[]]

// Inferring parts of tuple types

declare function foo<T extends string[], U>(...args: [...T, () => void]): T;

foo(() => {});  // []
foo('hello', 'world', () => {});  // ["hello", "world"]
foo('hello', 42, () => {});  // Error, number not assignable to string

// Inferring to a composite tuple type

function curry<T extends unknown[], U extends unknown[], R>(f: (...args: [...T, ...U]) => R, ...a: T) {
    return (...b: U) => f(...a, ...b);
}

const fn1 = (a: number, b: string, c: boolean, d: string[]) => 0;

const c0 = curry(fn1);  // (a: number, b: string, c: boolean, d: string[]) => number
const c1 = curry(fn1, 1);  // (b: string, c: boolean, d: string[]) => number
const c2 = curry(fn1, 1, 'abc');  // (c: boolean, d: string[]) => number
const c3 = curry(fn1, 1, 'abc', true);  // (d: string[]) => number
const c4 = curry(fn1, 1, 'abc', true, ['x', 'y']);  // () => number
```

## Structure and instantiation

The basic structure of a non-generic tuple type remains unchanged with this PR: Zero or more required elements, followed by zero or more optional elements, optionally followed by a rest element (for example `[A, B?, ...C[]]`). However, it is now possible to have _variadic elements_ anywhere in a tuple type, except following the optional rest element (for example `[A, ...T, B?, ...U, ...C[]]`).

A variadic elemement is a spread element of the form `...T`, where `T` is a generic type constrained to any array or tuple type (specifically, any type that is assignable to `readonly any[]`). Intuitively, a variadic element  `...T` is a placeholder that is replaced with one or more elements through generic type instantiation. Instantiation of a tuple type with a variadic element `...T` depends on the type argument provided for `T` as follows:

* When the type argument for `T` is a union type, the union is spread over the tuple type. For example, `[A, ...T, B]` instantiated with `X | Y | Z` as the type argument for `T` yields a union of instantiations of `[A, ...T, B]` with `X`, `Y` and `Z` as the type argument for `T` respectively.
* When the type argument is a tuple type, `T` is replaced with the elements of that tuple type. For example, `[A, ...T, B]` instantiated with `[X, Y]` as the type argument for `T` yields `[A, X, Y, B]`.
* When the type argument is an array type, `T` is replaced with a rest element. For example, `[A, ...T]` instantiated with `X[]` as the type argument for `T` yields `[A, ...X[]]`.
* When the type argument is another generic type, `T` is simply replaced with that type.
* When the type argument is `any`, `T` is replaced with `...any[]`.
* When the type argument is `never`, the entire result is `never`.

Instantiation of a generic tuple type includes _normalization_ to ensure the resulting tuple type follows the basic structure described above. Specifically:

* Optionality is removed from any optional elements that precede a required element. For example, `[A?, ...T, B?]` instantiated with `[X, Y?]` as the type argument for `T` yields `[A, X, Y?, B?]`.
* Rest elements absorb any following elements. For example, `[A, ...T, B]` instantiated with `X[]` as the type argument for `T` yields `[A, ...(X | B)[]]`.
* Tuples with only a single rest element are reduced to array types. For example, `[...X[]]` is reduced to simply `X[]`.

## Type relationships

Generally, a tuple type `S` is related to a tuple type `T` by pairwise relating elements of `S` to the elements of `T`. Variadic elements are processed as follows:

* A variadic element `...U` in `S` is related to a variadic element `...V` in `T` if `U` is related to `V`.
* A variadic element `...U` in `S` is related to a rest element `...X[]` in `T` if `U` is related to `X[]`.

Some examples:

```ts
function foo1<T extends unknown[], U extends T>(x: [string, ...unknown[]], y: [string, ...T], z: [string, ...U]) {
    x = y;  // Ok
    x = z;  // Ok
    y = x;  // Error
    y = z;  // Ok
    z = x;  // Error
    z = y;  // Error
}
```

Tuple types with single variadic elements have the following relations:

* `[...T]` is related to `T`.
* `T` is related to `readonly [...T]`.
* `T` is related to `[...T]` when `T` is constrained to a mutable array or tuple type.

Some examples:

```ts
function foo2<T extends readonly unknown[]>(t: T, m: [...T], r: readonly [...T]) {
    t = m;  // Ok
    t = r;  // Error
    m = t;  // Error
    m = r;  // Error
    r = t;  // Ok
    r = m;  // Ok
}
```

## Type inference

Inference between tuple types with the same structure (i.e. same number of elements and fixed, variadic, or rest kind matched to the same kind in each position), simply infers pairwise between the element types. For example, inference from `[string, ...Partial<S>, number?]` to `[string, ...T, number?]` infers `Partial<S>` for `T`.

Inference between tuple types `S` and `T` with different structure divides each tuple into a starting fixed part, a middle part, and an ending fixed part. Any one of these parts may be empty.

* The starting fixed parts of `S` and `T` consist of those elements in `S` and `T` that are fixed (i.e. neither variadic nor rest elements) in both types matching from the start of each type.

* If `T` contains at least one variadic element and `S` has no ending rest element, the ending fixed parts of `S` and `T` consist of those elements in `S` and `T` that are fixed in both types matching from the end of each type.

* If `T` contains at least one variadic element and `S` has an ending rest element, the ending fixed part of `T` consists of those elements in `T` that are fixed matching from the end of the type, and the ending fixed part of `S` is empty.

* If `T` contains no variadic elements, the ending fixed parts of `S` and `T` are empty.

* The middle parts of `S` and `T` are those elements in `S` and `T` that remain between the starting and ending fixed parts of the types respectively.

Inference then proceeds as follows:

* Pairwise infer between the elements in the starting parts.

* If the middle part of `S` is a single rest element, infer from that rest element to every element in the middle part of `T`.

* If the middle part of `T` is a single variadic or rest element, infer from a tuple consisting of the middle part of `S` to that variadic or rest element.

* If the middle part of `T` is exactly two variadic elements `...A` and `...B`, and an _implied arity_ exists for `A`, infer from a tuple consisting of the initial middle part of `S` to `A` and from a tuple consisting of the remaining middle part of `S` to `B`, where the length of the initial middle part corresponds to the implied arity for `A`.

* Pairwise infer between the elements in the ending parts, or infer from the rest element in `S` to the elements of the ending part of `T`.

In the context of inference for a call of a generic function with a rest parameter `R`, the implied arity for `R` is the number of rest arguments supplied for `R`. In all other contexts, a type parameter has no implied arity. For an example of inference involving an implied arity, see the `curry` function in the introduction.

Some examples:

```ts
type First<T extends readonly unknown[]> = T[0];
type DropFirst<T extends readonly unknown[]> = T extends readonly [any, ...infer U] ? U : [...T];

type Last<T extends readonly unknown[]> = T extends readonly [...infer _, infer U] ? U : undefined;
type DropLast<T extends readonly unknown[]> = T extends readonly [...infer U, any] ? U : [...T];

type T1 = First<[number, boolean, string]>;  // [number]
type T2 = DropFirst<[number, boolean, string]>;  // [boolean, string]
type T3 = Last<[number, boolean, string]>;  // [string]
type T4 = DropLast<[number, boolean, string]>;  // [number, boolean]
```

## Spreads in array literals

When an array literal has a tuple type, a spread of a value of a generic array-like type produces a variadic element. For example:

```ts
function foo3<T extends unknown[], U extends unknown[]>(t: [...T], u: [...U]) {
    return [1, ...t, 2, ...u, 3] as const;  // readonly [1, ...T, 2, ...U, 3]
}

const t = foo3(['hello'], [10, true]);  // readonly [1, string, 2, number, boolean, 3]
```

When the contextual type of an array literal is a tuple type, a tuple type is inferred for the array literal. The type `[...T]`, where `T` is an array-like type parameter, can conveniently be used to indicate a preference for inference of tuple types:

```ts
declare function ft1<T extends unknown[]>(t: T): T;
declare function ft2<T extends unknown[]>(t: T): readonly [...T];
declare function ft3<T extends unknown[]>(t: [...T]): T;
declare function ft4<T extends unknown[]>(t: [...T]): readonly [...T];

ft1(['hello', 42]);  // (string | number)[]
ft2(['hello', 42]);  // readonly (string | number)[]
ft3(['hello', 42]);  // [string, number]
ft4(['hello', 42]);  // readonly [string, number]
```

## Indexing and destructuring

Indexing and destructuring of generic tuple types appropriately recognizes fixed elements at the start of the tuple type. Beyond the fixed elements, the type is simply a union of the remaining element types.

```ts
function f1<T extends unknown[]>(t: [string, ...T], n: number) {
    const a = t[0];  // string
    const b = t[1];  // [string, ...T][1]
    const c = t[2];  // [string, ...T][2]
    const d = t[n];  // [string, ...T][number]
}

function f2<T extends unknown[]>(t: [string, ...T, number], n: number) {
    const a = t[0];  // string
    const b = t[1];  // [string, ...T, number][1]
    const c = t[2];  // [string, ...T, number][2]
    const d = t[n];  // [string, ...T, number][number]
}

function f3<T extends unknown[]>(t: [string, ...T]) {
    let [...ax] = t;  // [string, ...T]
    let [b1, ...bx] = t;  // string, [...T]
    let [c1, c2, ...cx] = t;  // string, [string, ...T][1], T[number][]
}

function f4<T extends unknown[]>(t: [string, ...T, number]) {
    let [...ax] = t;  // [string, ...T, number]
    let [b1, ...bx] = t;  // string, [...T, number]
    let [c1, c2, ...cx] = t;  // string, [string, ...T, number][1], (number | T[number])[]
}
```

## Rest parameters and spread arguments

Spread expressions with fixed length tuples are now appropriately flattened in argument lists. For example:

```ts
declare function fs1(a: number, b: string, c: boolean, ...d: number[]): void;

function fs2(t1: [number, string], t2: [boolean], a1: number[]) {
    fs1(1, 'abc', true, 42, 43, 44);
    fs1(...t1, true, 42, 43, 44);
    fs1(...t1, ...t2, 42, 43, 44);
    fs1(...t1, ...t2, ...a1);
    fs1(...t1);  // Error: Expected at least 3 arguments, but got 2
    fs1(...t1, 45);  // Error: Type '45' is not assignable to type 'boolean'
}
```

A rest parameter of a generic tuple type can be used to infer types from the middle part of argument lists. For example:

```ts
declare function fr1<T extends unknown[]>(x: number, ...args: [...T, number]): T;

function fr2<U extends unknown[]>(u: U) {
    fr1(1, 2);  // []
    fr1(1, 'hello', true, 2);  // [string, boolean]
    fr1(1, ...u, 'hi', 2);  // [...U, string]
    fr1(1);  // Error: Expected 2 arguments, but got 1
}
```

## Application of mapped types

When a mapped type is applied to a generic tuple type, non-variadic elements are eagerly mapped but variadic elements continue to be generic. Effectively, `M<[A, B?, ...T, ...C[]]` is resolved as `[...M<[A]>, ...M<[B?]>, ...M<T>, ...M<C[]>]`. For example:

```ts
type TP1<T extends unknown[]> = Partial<[string, ...T, number]>;  // [string?, ...Partial<T>, number?]
```

Fixes #5453.
Fixes #26113.

Artazor commented:
Wow! Just Wow! Waited for it since original proposal #5453

RyanCavanaugh commented:
I'll do the usual incantations
@typescript-bot perf test this
@typescript-bot user test this
@typescript-bot test this
@typescript-bot run dt
@typescript-bot pack this

RyanCavanaugh commented:
This appears to be a new crash identified:

https://github.com/typescript-bot/TypeScript/pull/43/files#diff-0e62d568d28b82cee9e66b2b8fafd290R13

aminpaks commented:
Is this what we expect from `t1`?
```ts
function foo<T extends unknown[]>(t: [...T]) {
    return t[0];
}

const t1 = foo([false, 'check', 'base']); // string | boolean
```

[Playground link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=1&ssc=1&pln=6&pc=1#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdFwBrMOAdzAG0BdAPgAooAuRegOj40YBKRAG8AUIkmIATtighpSKPQAMjANxiAvmLEQEJKIigBGRAF5kqVvWABDADYlsAGkQByCAAtsECu7d3ACM7Z3chdUQAeijEQ2kYMABzRAAfRCDUB2w7MDEgA)

DanielRosenwasser commented:
There's also this weirdness around reporting tuple properties as "optional" in error messages. https://github.com/typescript-bot/TypeScript/pull/43/files#diff-a63ec2ec012ff897892e9da0bb5df6d6R188

mohsen1 commented:
@aminpaks you need a tuple not an array. [Playground link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdFwBrMOAdzAG0BdAPgAooAuRAJ2wEMiCADYBPRPQB0UjIwCUiAN4AoRKp7YoIbkij0ADIwDcSgL5KlEBCSiIoARkQBeZKlb1gfISWwAaRAHIIAAtsCAp-P38AIz5vf0ZEWMRLMGtZQ0QAekzEa24YMABzRAAfRCjUIX4wIA)

IllusionMH commented:
Is this PR expected to enable typing `curry` function from #5453 ?
[Attempt in playground](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7#code/GYVwdgxgLglg9mABBEAnVBPAPAWAFCKGIAqiApgB5RlgAmAzouANZhwDuYA2gLoA0+IogCq5KjQZMwrDt36CiANXwA+ABQLCwAFyI1AOkNR6urof3E+ic8J4BKRAF4ViRQIJFzAQ13F8d3QNDACNdYQdnV0QAb01EVDIoNCQg-WCIl2BUrytzdIBufABffAgEeihEGDAABxBKxz0fRDAQAFtgslQrUMQK1GqAcysIXWC4OAAbMi8wK1pdegwOqYzELhzEYJH5nkQvRjKwCsK8UvLK9jhUZkZGlHQMNWq6qCsARisAcneAJgBmL5WKCoEBkKwAZWW40majsdlORwq5HQ1ycyDQmGetXqH2+f0BdjUILBkOhUzhCPw+AA9DT6F42mREOwYFAABbWQzNMyGYh7Wa0LlpUw2HhAA)

ahejlsberg commented:
@aminpaks Yes, I think it is reasonable to expect the inferred type of `t[0]` to be `[...T][0]`, which would then resolve to `boolean` in the call to `foo`. It's an easy change.

@IllusionMH Yeah, the final cherry on top is to handle partial `curry` or `bind`. I haven't gotten to it yet, but I'm intending to. It's definitely possible to infer the cut point from the number of arguments passed.

ddprrt commented:
@IllusionMH Currying works nice, see [Playground](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=1&ssc=1&pln=26&pc=42#code/C4TwDgpgBAYgrgOygXigCgHReAZwFxQCGCIAlCgHxEkCwAUKJFAJIIBmEATjAJac7AAPDCrJ6UCbCgQAHsAgIAJjnRYMuAgG01PdlygAVALrlkVYiCgB+cZIPS5C5VG1ZdHTlACqAGig69TwA1I2tvKAIECAA3fUiYrgBuenpGaFYPAGUIAGMAeyVhP3tZeSUVRABrBDyAdwRNI1FbCRgHMudMbHwXNQM-NSh3fS8TSiHAqCCwrwioKNjOZIZwdMCAJQgcOAAbIREUFqlSpxUu9R7XDGHPfv8sCY9vMbNH-WmrKbmFpJS6NkQOWAPAKUBycE4nBAgiOJUc5VgiB8R14-GA7VOUCqNXqjWRdEkUGy+SUGIRGS4xIKikEd1RAgo9AoaCObAI-SOakIBHpwHopAI5wARgQqUpTFQKZxNts9rSqABvI6cCDACFIYUSqBsc6EAZYIWkZYAX3oJIEEzAcHRqDQ3PmcAAtkKuH4RVABJxdABzPw5AhCvJ5HYQYh+RQEHAgZ3BrWaPVQIV+8OhQgqc3AZZmgoW2p5TiVFSocGQkBoXRW4B+ACMfgA5NWAEwAZjrfmAnDgED8mWjgZ2aFIpEHywzUDzBZwAEEdjg8igwRCoeWEJWa-Wm63hx2uz2+8HB0a-gB6Y84QiO6C1HjAAAW9ww9quxmoigf7quoyAA)

The trick is to bind the function arguments so they become value-types, then infer the missing ones

```typescript
type Fun = (...ts: any) => any
type InferSecond<F, T extends unknown[]> =
    F extends (...ts: [...T, ... infer U]) => infer V ? U : never;
type InferResult<F> =
    F extends (...ts: any[]) => infer V ? V : never;

function curry<
    T extends Fun,
    First extends unknown[],
    Second extends InferSecond<T, First>
>(
    f: T,
    ...a: First
): (...b: Second) => InferResult<T> {
    return (...b) => f(...a, ...b);
}
```

jwbay commented:
Would this also close #26113?

AlexAegis commented:
I haven't seen an example on generics, are those supported too?

Will I have able to define a function signature, that when I implement I get something like this?

```ts
function foo<...T>(...t: ...T[]) { ... }

foo<string, number, string>('a', 2, 'b'); // this should be callable exactly like this
```

The important part here would be to not define the generics on foo as a tuple, to not break existing code when migrating from overly verbose workarounds like this:

https://github.com/ReactiveX/rxjs/blob/master/src/internal/observable/combineLatest.ts#L63

RyanCavanaugh commented:
Internal code suite shows only diffs in error messages

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -17926,7 +18140,7 @@ namespace ts {
         }
 
-        function isArrayType(type: Type): boolean {
+        function isArrayType(type: Type): type is TypeReference {
```
This shouldn't be a type guard (we haven't done #15048)

RyanCavanaugh replied:
Or rather the return type could be `type is (TypeReference & (something else))}` to avoid incorrect logic in `else` branches give a `TypeReference | Foo`

ahejlsberg replied:
In principle I agree, but the type guard does save several type casts. I suppose we could introduce an `ArrayTypeReference` and use that instead.

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -18046,8 +18260,16 @@ namespace ts {
+            return isTupleType(type) && !!(type.target.combinedFlags & ElementFlags.Variadic);
+        }
+
+        function isSingleElementGenericTupleType(type: Type): type is TupleTypeReference {
```
Same re: type guardness

ahejlsberg replied:
Hmm, not crazy about removing these or introducing types like `SingleElementGenericTupleTypeReference` that we don't otherwise need.

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -18056,7 +18278,30 @@ namespace ts {
 
         function getLengthOfTupleType(type: TupleTypeReference) {
-            return getTypeReferenceArity(type) - (type.target.hasRestElement ? 1 : 0);
+            return type.target.fixedLength;
```
Seems like this function should either be inlined and removed, or other places where `type.target.fixedLength` is accessed should call it instead (e.g. line 18272 above)

ahejlsberg replied:
Agree.

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -18056,7 +18278,30 @@ namespace ts {
+            return undefined;
+        }
+
+        function isMatchingTupleTypes(t1: TupleTypeReference, t2: TupleTypeReference) {
```
This doesn't have the very similar semantics to other `isMatching` functions we have; I think a comment is warranted or maybe a name change. I was expecting this function to compare the types themselves, not just the overall "binding shape" if you will

ahejlsberg replied:
Maybe `isTupleTypeStructureMatching`?

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -25646,19 +25895,27 @@ namespace ts {
+                const effectiveArgs = args.slice(0, spreadIndex);
+                for (let i = spreadIndex; i < args.length; i++) {
+                    const arg = args[i];
+                    // We can call checkExpressionCached because spread expressions never have a contextual type.
```
👀 This surprised me but I can't come up with a reason why it's wrong to not flow the contextual types in, since you should never need to have a literal like that inside a spread position

ahejlsberg replied:
Right, and BTW this is nothing new. Even before this PR spread expressions didn't have contextual types.

aminpaks commented:
Just ran into this one too. I was expecting the same result
```ts
function foo<T extends unknown[]>(...p: T) {
    return p[0];
}

declare const params: [number, boolean];

const v2 = foo(...params); // unknown
```

[Playground link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=1&ssc=1&pln=7&pc=38#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdFwBrMOAdzAG0BdAPgAoA6TgBwC5EMAlIgDeAKEQTEAJ2xQQUpF3oAGRgG5RAX1Gii2CABsAhjMQQEJKIi4mjAWxJ96YEHYBG2KQBpEb1AewjMHUdczBLRAA3ACZEAF5kVA5uWwcBNUQAekzyMCpaMCA)

ahejlsberg commented:
> Would this also close #26113?

Yes indeed.

IllusionMH commented:
@ddprrt thanks, that's an interesting solution :+1:

@ahejlsberg for function first it's clear how to infer parameters that are expected later (for single or 2 calls as in example from @ddprrt)

But is it possible to have "function last" as in [`dependentExpected` example in Playground](https://www.staging-typescript.org/play?ts=4.0.0-pr-39094-7#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXy1VAAcQiyMAeAFXhAA8MzgBneNAa1RwHdUBtALoAaeACUAfAAoAdHNgBzFgC54-OTOqjZ8mEtVRUATyEBKeAF4J4wadX8tNgNwAoF4RLMKUqeavwAb3gAX1MneHgAeki1IVEANxwsYEF3clJyVAwpAHIACxAICBwc0RyeHBgIYFL4X0trINDwqJj+fhYMGEIFUU7u1AUReETk1LTPTOz8wuLagBYAJm0-awBGMNbY-p7RVGQAWwAjEBhh-ePT1I8QDNAs3IKiktEl7ShVHcHV+ABCKE20W2XV28AuJzOoiOOBwEBAhnGLlAkFgCBQ6GweHgk3uVFoDCYRDYnG4fDi8AAqnRGMw2I5JDoZIoVGoNI5Gcz7BoKbYGjY7Gp6YJXEjbl4HvV-E1Nlt2sNRilRXdvDNnrVypVqrVJY0QjKge0vr14Eb5UlFUrxdMnnNXss6gC+QDwgaOiDBntDhDhiTeKhBPBALwbgCI9y1TR6zF7wN4OqE-fhQKEBqBsMB4TqAtpugbG8FXURwKDAPAQIxqX18UQV-0BkNh3ERtV295Q0RgeOJ+BHNvJ1PpjCZ4E5z2XSHwQvF1Cl8uoLh+quz0moBdzviCWuhtzI6BwJBoTC4fA4igAUXopEwIGAlAZHxN7uNR1UeZgbafdQ0nLURpH3p+YgFQ0H1-fNnC3MUpjPC8mGARtbWje0dUCPUWldH8wS9UCFXGY8sig8AYLgqMYykR1-GdWVs1BF9hlNetT3PAiryIhZEM7I4OyTeAU3gNNUAzF0s3QmiC3hSdpyoj0MNHVJcIwfDL1g1V4JI9i204rse24vt+IHQT4BPGAYEqIA) ?

It's clear that function can be used as last argument, but as soon as `T` used in arguments for last function inference of first two arguments "is broken", and if `U extends T` used instead if will be `unknown[]`.

HerringtonDarkholme commented:
Inspired by the brilliant ideas from @IllusionMH and @ddprrt, I almost rolled out a generic `compose` but something is still missing.  Any feedback is welcome.

Specifically, 

```typescript
type Query<K, T> = K extends keyof T ? T[K] : never

// Zip loses tuple type
type Zip<As extends unknown[], Bs extends unknown[]> = {
    [key in keyof As]: (a: As[key]) => Query<key, Bs>
}
type Compose<R extends unknown[]> = Zip<Lead<R>, Tail<R>>

// Head, Last, Lead, Tail definition omitted here 
declare function compose<R extends unknown[]>(...fns: Compose<R>): Fn<Head<R>, Last<R>>
                                                                                               ↑~~~~ Compose is not tuple type
```
Since Zip's type parameter must be array like, would it be possible to infer `Zip`'s return type as array like as well? 

[Playground Link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=1&ssc=1&pln=22&pc=23#)

amitport commented:
this is great, I want to ask about this part though:

```ts
declare function ft1<T extends unknown[]>(t: T): T;
declare function ft3<T extends unknown[]>(t: [...T]): T;

ft1(['hello', 42]);  // (string | number)[]
ft3(['hello', 42]);  // [string, number]
```

looking only on the template param (`T extends unknown[]`) and declared output type (`T`) it seems to me very confusing that it generates a different output type.
why can't it be something like: (?)
```ts
declare function ft1<T extends unknown[]>(t: T): T;
declare function ft3<T extends unknown[]>(t: [...T]): T;
declare function ft5<T extends unknown[]>(t: [...T]): [...T];

ft1(['hello', 42]);  // (string | number)[]
ft3(['hello', 42]);  // (string | number)[]
ft5(['hello', 42]);  // [string, number]
```

DanielRosenwasser commented:
Noticing that some of the names are lost in this case:

```ts
type CallbackStyle<Args extends any[], Return> =
  (..._: [...args: Args, cb: (err: NodeJS.ErrnoException | null, result: Return) => void]) => void


type Foo = CallbackStyle<[number, string], string>
```

In that case, `Foo` is displayed as

```ts
type Foo = (__0: number, __1: string, __2: (err: NodeJS.ErrnoException | null, result: string) => void) => void
```

[Link here](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=6&ssc=1&pln=1&pc=1#code/C4TwDgpgBAwghgGwQIzgYwNYGVQIgHgEEAnAcwGcoIAPYCAOwBNK56QBtAXQBooAlCMACuxegD4oAXgCwAKChQAFADpVAfQBcUdquVwy5LSQq80yLYojFiWgHIB7RhABSWZQFFr9e++poIYMAAlvb0UAA+UPRCSLzEEOQxwFoCwqIAlFISAG72QYycmZI5eYxy5bKgkFAAYvb2UrCIKOjYuATs0QC2yFa85MDEQfSkPFADQyNiFUFdYPbEwFAAVFBwlABmm8T2XVAARFv7ANwVTmgI+tAbQvRowaFQYDtdQeRBGyBEBlS0DMxrNhcXgCRIIYBiRRyBQbehaeBIVCYHAgPDfEz8BJJaaydIWXT6ChGAxFCQABRebwIoOxp1kcjwSw2jWeuzeHxAii2ynicEYNSCeHSQA)

TylorS commented:
Firstly, this is an amazing feature. It will help with many abstractions in the TS codebases I've worked in.

It'd be *amazing* if we could type the yields of a generator using these variadic tuple types. Any plans of this or general interest (I'll happily open another issue)? 

Myself and others have been collaborating on creating an algebraic effects abstraction based on generator functions, and the union of yields is much harder to work with. It'd be very convenient to be able to used mapped types over these values, which can be used for aggregating the requirements.

Thanks for your time!

ahejlsberg commented:
@HerringtonDarkholme Your playground link doesn't work so can't see the full example.

ahejlsberg commented:
@DanielRosenwasser You get named tuple members only when all elements have names. For example, it works for this instantiation:

```ts
type Foo = CallbackStyle<[a: number, b: string], string>
```

ahejlsberg commented:
> But is it possible to have "function last" as in dependentExpected example in Playground?

@IllusionMH That should have worked, and now does with my latest commit.

DanielRosenwasser commented:
Duh - sorry, it was late by me

sandersn commented:
Here's the smallest crash I've got in VS Code:

```ts
declare function crash(...args: unknown[]): any;
export function gift(us: unknown[]) {
  crash(...us);
}
```

The actual crash is in vscode/src//vs/editor/standalone/browser/standaloneCodeEditor.ts:291, when trying to resolve this call:

`_descriptor.run(this, ...args)`

When checking assignability, the two types &mdash; *I think* for the spread/rest &mdash; are missing symbols, which then crashes inside resolveStructuredTypes.

I'm going to try to get a standalone repro and then try debugging again. But maybe this is enough to go on.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -25684,7 +25939,7 @@ namespace ts {
-                const spreadType = getSpreadArgumentType(args, argCount, args.length, restType, /*context*/ undefined);
+                const spreadType = getSpreadArgumentType(args, argCount, args.length, restType, /*context*/ undefined, checkMode);
                 const errorNode = reportErrors ? argCount < args.length ? args[argCount] : node : undefined;
                 if (!checkTypeRelatedTo(spreadType, restType, relation, errorNode, headMessage, /*containingMessageChain*/ undefined, errorOutputContainer)) {
```
a bit more detail from debugging the VS Code crash again: 

```ts
declare function crash(...args: unknown[]): any;
export function gift(us: unknown[]) {
  crash(...us);
}
```

at this point, spreadType is the reference object type for `us`, whose target is an anonymous object type. Neither type has a symbol. When resolveStructuredTypeMembers is called on spreadType inside of checkTypeRelatedTo (relating `us` to `args`), resolveTypeReferenceMembers calls resolveDeclaredMembers spreadType.target. resolveDeclaredMembers assumes that spreadType.target has a symbol, which it does not.

sandersn replied:
The only reason we're in this block is that `restType` is defined. A standalone run has `restType` undefined because it's not an array type. That must mean that VS Code's compile messes up `globalArrayType` in some way -- changing the program to use `Array<unknown>` avoids the crash:

```ts
declare function crash(...args: Array<unknown>): any;
export function gift(us: unknown[]) {
  crash(...us);
}
```

ahejlsberg replied:
Ah, maybe it only happens when the global array type couldn't be located. We then default it to an empty object type--which has no symbol.

ahejlsberg replied:
Yeah, that appears to be the case. When I compile the repro with `tsc -nolib` it does indeed crash.

sandersn replied:
That's what I confirmed in the debugger. Is the existing code in the checker resilient to missing the Array type?

ahejlsberg replied:
Now fixed.

isiahmeadows commented:
Is it me, or have you all not yet implemented circular type aliases for it?

[Playground Link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7#code/C4TwDgpgBAMghgZ2AHgCpQgD2BAdgEwSgCcI58B7XAGxCgFdcBrXCgd1wG0BdAPigC8AWABQUcVHRYcBIqXJVaUTgDo1AS1wAzCMSgB9ADRRNOvQFVuUAPxRzUAFwMCELZoj4A3KNGacxLTgAY2hzXDhiEAAxRiDgdSo0YwAlfgBvKAAKBAp6YhCnVABKJ2TPKABfHxFQSCgABXVIeoi4AFsEJKhkjGw8QmcWdi4+QVEJbt6ZAflKGjpOU107YzUVJb0ANW5xiVtOMIjo2PjE1GNzXlW1RubWjuRzY03eHbEJJx5vEVF8CCDqBFoFoTglcFAwE0IF0etJ+kRGEMODxeJk1lpcAgnLcIC1iO1OudurwSnZwpEYrg4mCuvAkMhUrxPEA)

ahejlsberg commented:
@typescript-bot user test this

HerringtonDarkholme commented:
Sorry for the inconvenience. Here is the example for compose

```ts
type Query<K, T> = K extends keyof T ? T[K] : never
type Zip<As extends unknown[], Bs extends unknown[]> = {
    [key in keyof As]: (a: As[key]) => Query<key, Bs>
}
type Head<R extends unknown[]> = R[0]
type Tail<R extends unknown[]> = R extends [any, ...infer T] ? T : []
type Lead<R extends unknown[]> = R extends [...infer L, any] ? L : []
type Last<R extends unknown[]> = R extends [...Lead<R>, infer L] ? L : never

type Compose<R extends unknown[]> = Zip<Lead<R>, Tail<R>>

declare function compose<R extends unknown[]>(...fns: Compose<R>): Fn<Head<R>, Last<R>>
//                                                    ↑~~~~ Compose is not tuple type
```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```


</details>


**Playground Link:** [Provided](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=1&ssc=1&pln=13&pc=86#)

pirix-gh commented:
Thanks for this amazing PR. So I was testing to replace my `Concat` implementation on the ts-toolbelt

```ts
export type Concat<L extends List, L1 extends List> =
    [...L, ...L1]
```

instead of the old

```ts
export type Concat<L extends List, L1 extends List> =
    _Reverse<_Reverse<L>, L1>
```

(https://github.com/pirix-gh/ts-toolbelt/blob/master/src/List/Reverse.ts)

It turns out that the memory used for my tests doubled from 150MB to 350MB, with a slight performance increase of 1s. 
It would be great if we could lower the memory consumption :smile_cat:

treybrisbane commented:
Similarly to @isiahmeadows comment above, this attempt at a `pipe` function also doesn't appear to work: [[playground link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=3&ssc=1&pln=3&pc=10#code/C4TwDgpgBAMghgZ2AHgCpQgD2BAdgEwSgCcI58B7XAGxCgFdcBrXCgd1wG0BdAPigC8UdFhwEipclVpROAOgUBLXADMIxKAH0ANFGVqNAVW5QA-FENQAXAwIQVyiPgDcAWABQH5TmIq4AY2hDXDhiEAAxRn9gRSo0XQAlfgBvKAAKBAp6YkCbVABKGwTnKABfDw9QSCgABUVIGtC4AFsEeKgEjGw8QlsWdi4+QShkjyhxqAAGGx5tMYmARiKusV7JSho6Tn11C10FOR2NADUTU3nxzmDQiKiYuNRdQ159hTqGptbkQ11j3hMrBcoLgIAA3dQeUqcTqiHoSMgbGScRj9DhnKALaxTbhuTzufAQfzUULQfxUJBQMD1CA2dow7riPqsDg8XhpA4qXAIGzvCCNYgtNqPDq8fKCfjXMKRXDRWK4drwJDIJK8XEeAD06qgAFpdXr9QbDUbjSbTSaKvjCcTSFAyVzgCQsgQbGlcDZcPRmgAjdRigT8D3e9Rqy1Ekm28kOyj0L3UGnpN3Az0+4h+gPJ4MWglhm12ilIYjKADmihUIBdcBsKOZuDTUALxZDmqgAGUABZZaj4YEUB3qYgUYgeKmQNJAgeMfBzdwTKDR2MQaezhu4Etl6f5NIAFgATHIdwBWfIeIA)]

IllusionMH commented:
> That _[function last]_ should have worked, and now does with my latest commit.

Thank you! Works as expected with latest commits (not in Playground yet)

```ts
type SlectorFor<T> = { __resultType: T; };
type UnwrapSelectors<T extends SlectorFor<unknown>[]> = {
    [K in keyof T]: T[K] extends SlectorFor<unknown> ? T[K]['__resultType'] : never
};

declare const numSelecotor: SlectorFor<number>;
declare const strSelector: SlectorFor<string>;
declare const boolSelector: SlectorFor<boolean>;

declare function createSelector<T extends SlectorFor<unknown>[], R>
    (...args: [...T, (...args: UnwrapSelectors<[...T]>) => R]): SlectorFor<R>;

const res1: SlectorFor<string> = createSelector(numSelecotor, strSelector, (a, b) => b.repeat(a));
const res2: SlectorFor<number> = createSelector(numSelecotor, strSelector, (a) => a * 5);
const error: SlectorFor<string> = createSelector(numSelecotor, (a, b) => b.repeat(a));
```

uhyo commented:
Thank you for this very exciting feature! Really looking forward for TS 4.0.
Especially, with the "function last" pattern, we can define a `pipe` function below and type inference fully works.

```ts
function pipe<Values extends readonly any[]>(
    ...args: [...Values, (...values: Values) => void]
) {
    const callback = args[args.length - 1] as (...values: Values) => void
    const values = args.slice(0, -1) as unknown as Values
    callback(...values)
}

pipe("foo", 123, true, (a, b, c) => {
    // (parameter) a: string
    // (parameter) b: number
    // (parameter) c: boolean
})
```

**However**, I'm feeling that it is very hard to correctly deal with the "absorb" behavior. For example, the above `pipe` function breaks without compile errors when `Values` is instantiated with a variable-length array type:

```ts
pipe<number[]>(1, 2, 3, 4) // runtime error!

const a = [1, 2, 3];
pipe(...arr); // runtime error!
```

So it would be amazing to have a way to allow only fixed-length tuples.

ahejlsberg commented:
@pirix-gh Not sure what the increased memory consumption is, but I'm fairly sure it is unrelated to this PR. I'm seeing slightly lower memory consumption and slightly lower check times (Node.js v12.14.0 on Windows):

```
// TS 3.9.5

Files:              314
Lines:            41034
Nodes:           184648
Identifiers:      61841
Symbols:         189102
Types:           105441
Instantiations:  628013
Memory used:    188865K
I/O read:         0.03s
I/O write:        0.00s
Parse time:       0.47s
Bind time:        0.22s
Check time:       7.85s
Emit time:        0.00s
Total time:       8.54s

// This PR

Files:              315
Lines:            41371
Nodes:           186320
Identifiers:      62017
Symbols:         133698
Types:            92117
Instantiations:  435318
Memory used:    181397K
I/O read:         0.04s
I/O write:        0.00s
Parse time:       0.59s
Bind time:        0.21s
Check time:       7.16s
Emit time:        0.00s
Total time:       7.96s

// This PR with Concat as [...T, ...U]

Files:              315
Lines:            41372
Nodes:           186327
Identifiers:      62017
Symbols:         133229
Types:            91766
Instantiations:  434115
Memory used:    176498K
I/O read:         0.03s
I/O write:        0.00s
Parse time:       0.59s
Bind time:        0.21s
Check time:       7.11s
Emit time:        0.00s
Total time:       7.92s
```

ahejlsberg commented:
@uhyo Yes, the "absorb" behavior is not ideal for that scenario. To properly handle it we would need to extend the basic structure of a tuple from `[fixed-part, ...variable-part]` to `[starting-fixed-part, ...variable-part, ending-fixed-part]` and then update all of our reasoning in the compiler accordingly. It's not a trivial change, but it might be worthwhile.

pirix-gh commented:
@ahejlsberg Sorry about that, yes you are right. Not sure why tsc was showing the doubled memory consumption (consitently). I did the test again, it all looks right:

```ts
# With this PR
Files:              315
Lines:            41336
Nodes:           185024
Identifiers:      61973
Symbols:         111933
Types:            77121
Instantiations:  365661
Memory used:    152062K
I/O read:         0.02s
I/O write:        0.00s
Parse time:       0.83s
Bind time:        0.40s
Check time:      15.64s
Emit time:        0.00s
Total time:      16.86s
```

Thanks. Cheers!

bbrzoska commented:
> To properly handle it we would need to extend the basic structure of a tuple from `[fixed-part, ...variable-part]` to `[starting-fixed-part, ...variable-part, ending-fixed-part]` and then update all of our reasoning in the compiler accordingly. It's not a trivial change, but it might be worthwhile.

@ahejlsberg would this essentially mean implementing support for #6755?

ahejlsberg commented:
With the latest commit we can now infer to tuples of the form `[...T, ...U]`, provided `T` also occurs as the type of a rest parameter. I such cases we compute an _implied arity_ from the rest argument list and use that to determine the "cut point" between `...T` and `...U`.

```ts
function curry<T extends unknown[], U extends unknown[], R>(f: (...args: [...T, ...U]) => R, ...a: T) {
    return (...b: U) => f(...a, ...b);
}

const fn1 = (a: number, b: string, c: boolean, d: string[]) => 0;

const c0 = curry(fn1);  // (a: number, b: string, c: boolean, d: string[]) => number
const c1 = curry(fn1, 1);  // (b: string, c: boolean, d: string[]) => number
const c2 = curry(fn1, 1, 'abc');  // (c: boolean, d: string[]) => number
const c3 = curry(fn1, 1, 'abc', true);  // (d: string[]) => number
const c4 = curry(fn1, 1, 'abc', true, ['x', 'y']);  // () => number

const fn2 = (x: number, b: boolean, ...args: string[]) => 0;

const c10 = curry(fn2);  // (x: number, b: boolean, ...args: string[]) => number
const c11 = curry(fn2, 1);  // (b: boolean, ...args: string[]) => number
const c12 = curry(fn2, 1, true);  // (...args: string[]) => number
const c13 = curry(fn2, 1, true, 'abc', 'def');  // (...args: string[]) => number

const fn3 = (...args: string[]) => 0;

const c20 = curry(fn3);  // (...args: string[]) => number
const c21 = curry(fn3, 'abc', 'def');  // (...args: string[]) => number
const c22 = curry(fn3, ...sa);  // (...args: string[]) => number
```

ahejlsberg commented:
> would this essentially mean having support for #6755 as a side effect?

@bbrzoska No, it wouldn't do repeated patterns.

filipesabella commented:
I still don't quite get why the [playground link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=3&ssc=1&pln=3&pc=10#code/C4TwDgpgBAMghgZ2AHgCpQgD2BAdgEwSgCcI58B7XAGxCgFdcBrXCgd1wG0BdAPigC8UdFhwEipclVpROAOgUBLXADMIxKAH0ANFGVqNAVW5QA-FENQAXAwIQVyiPgDcAWABQH5TmIq4AY2hDXDhiEAAxRn9gRSo0XQAlfgBvKAAKBAp6YkCbVABKGwTnKABfDw9QSCgABUVIGtC4AFsEeKgEjGw8QlsWdi4+QShkjyhxqAAGGx5tMYmARiKusV7JSho6Tn11C10FOR2NADUTU3nxzmDQiKiYuNRdQ159hTqGptbkQ11j3hMrBcoLgIAA3dQeUqcTqiHoSMgbGScRj9DhnKALaxTbhuTzufAQfzUULQfxUJBQMD1CA2dow7riPqsDg8XhpA4qXAIGzvCCNYgtNqPDq8fKCfjXMKRXDRWK4drwJDIJK8XEeAD06qgAFpdXr9QbDUbjSbTSaKvjCcTSFAyVzgCQsgQbGlcDZcPRmgAjdRigT8D3e9Rqy1Ekm28kOyj0L3UGnpN3Az0+4h+gPJ4MWglhm12ilIYjKADmihUIBdcBsKOZuDTUALxZDmqgAGUABZZaj4YEUB3qYgUYgeKmQNJAgeMfBzdwTKDR2MQaezhu4Etl6f5NIAFgATHIdwBWfIeIA) @treybrisbane posted doesn't compile.

maxrothman commented:
I know this PR is still in progress, but I happened to be playing around with type-level tuple concatenation and noticed some issues with recursive and conditional types.

```typescript
type Concat<Arr extends any[][]> =   // Type alias 'Concat' circularly references itself.
  Arr extends [] ? [] :
  Arr extends [infer First, ...infer Rest] ? [...First, Concat<Rest>] :  // A rest element type must be an array type.
                                                                         // Type 'Concat' is not generic.
  never

type Concat2<Arr extends any[][]> =
  Arr extends [] ? [] :
  Arr extends [infer First extends any[], ...infer Rest] ? [...First, Concat<Rest>] :  // Exported type alias 'Concat2' has or is using private name ''.
                                                                                       // Exported type alias 'Concat2' has or is using private name ''.
                                                                                       // '?' expected.
                                                                                       // A rest element type must be an array type
                                                                                       // Type 'Concat' is not generic.
  never
```

Neither of these compiles. I'm not 100% confident it's not my mistake, but I figured I'd raise it just in case! [Playground link](https://www.typescriptlang.org/play/?ts=4.0.0-pr-39094-7&ssl=1&ssc=1&pln=10&pc=1#code/C4TwDgpgBAwg9gOwMYENgB4CCAnbUIAewECAJgM5QoIgDaAugwHxQC8AUFFDnocWZQZQA-FCEAuTt1z4iJCmICWCAGYQ8AMUXZywADRQAdMeVq8AJQi76IsccNad+2IlQZLupjfFQpCCABu6uzsoJAuyGgATFgyfPKU1HSM9CwcXDyy-ApCohJSmfECSqrqUI66WQlUNAwG9qZlHsA2efYVzvCR7lbAXlA+foHBQA)

ahejlsberg commented:
Regarding the various attempts to type `pipe` functions using variadic tuple types in conjunction with mapped and/or recursive types, it's not going to work without the introduction of new concepts in type inference. Consider:

```ts
type Fn<A, B> = (a: A) => B;

type Zip<A extends unknown[], B extends unknown[]> = {
    [K in keyof A]: Fn<A[K], B[K & keyof B]>;
};

type Head<R extends unknown[]> = R[0];
type Tail<R extends unknown[]> = R extends [any, ...infer T] ? T : [];
type Lead<R extends unknown[]> = R extends [...infer L, any] ? L : [];
type Last<R extends unknown[]> = R extends [...Lead<R>, infer L] ? L : never;

type Compose<R extends unknown[]> = Zip<Lead<R>, Tail<R>>;

declare function compose1<R extends unknown[]>(...fns: Compose<R>): Fn<Head<R>, Last<R>>;

declare function compose2<A, B, C>(...fns: Compose<[A, B, C]>): Fn<A, C>;

declare function toInt(s: string): number;
declare function double(n: number): number;

compose1(toInt, double);  // Error
compose2(toInt, double);  // Ok
```

The `compose2` example works because we first introduce a set of type parameters for which we then perform type inference. In `compose1` we have only one type parameter, `R`, yet we really want to perform type inference for a set of three type parameters, `R[0]`, `R[1]`, and `R[2]`. Yet it is not clear how we would infer _that_. I suppose we'd need some new way of indicating that we want to treat a type parameter as a tuple of individual inference slots, and some way of inferring the arity of that tuple. Definitely beyond the scope of this PR.

treybrisbane commented:
@ahejlsberg Well that's unfortunate. Thanks for the explanation. 🙂

jcalz commented:
Usually for things like this I try to see if an F-bounded constraint will work:

```ts
type Last<T extends any[]> = T extends [...infer _, infer L] ? L : never;
type Init<T extends any[]> = T extends [...infer I, any] ? I : never;
type Pipe<T extends ((x: any) => any)[]> = Init<[(x: any) => any, ...{
    [K in keyof T]: T[K] extends (x: any) => infer R ? (x: R) => any : never
}]>
declare function pipe<T extends ((x: any) => any)[] & Pipe<T>>(...f: T):
    (x: (T[0] extends (x: infer A) => any ? A : never)) =>
        Last<T> extends (x: any) => infer R ? R : never;

declare function toInt(s: string): number;
declare function double(n: number): number;

pipe(toInt, double);  // okay
pipe(double, toInt); // error in weird place due to #28505
```
🤷‍♂️

[Playground link](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-39094-7&ssl=14&ssc=1&pln=18&pc=1#code/LAKA9GAEC0t-DFOS1LSgC4E8AOBTSAGQEMBnTAHgBVJ8APTfAOwBMzITnsBtAXQB8kALyRaDJmw48AdHICWzAGb4ATpAD6AGkiKV6on0gB+YpABckZvgBuagNxY8hAJLN5VcYxbtO3fkKiXpK+sgrKapAuOlzYRqYuFla2Dk4EkAAK8gQ0dN5SkAAUhfSWsQCUIkIVASJR7lQ8JWXclcLV3DpyMgDeoJADkDwA0rrMkADW+NgA9kpifJbUI0YSPhzNfthtQnqRAEomRaWQ+ztbSdZ2qqAAvoKgrPgAxgA2JKqESgCuzM+Y8hm41w2XwuTWBWKJwqVS25X4kAAZJlQTQBAJCt0lEtyuZ+oNNoVlgAGVb5XybPbqACC51iR2plxSqnKO3xg0GpAoaLyIQ20NasKppyOh0sV1SIAwICebw+X1+-0B40wMzcmEKZEsFFUigA5rirN8ALYAI0lsven0gPz+AKBkFYM2+pte+EKzHFJvNLK9ZslIIIhVV6p0TpdbvK9gGEEgMwmJGwoED7vDrvwOhDzEwUcgsbUqhm6kUkAA7vh5KpWJBcO9noRWN9CKrIABiABMAA4AKzE7ugIA)

maxrothman commented:
@ahejlsberg is the inference issue you mentioned related to the problem with recursive types, or is that a separate issue? Here's a more minimal example ([playground link](https://www.typescriptlang.org/play/?ts=4.0.0-pr-39094-7&ssl=14&ssc=1&pln=14&pc=79#code/C4TwDgpgBAShYBsCGBjCAeA8ggJgGigDkIB3AgQSggA9gIA7HAZyiXpAG0BdAPigF4AUFCiUadRi25QA-FGkAuYaKq0GzednxQAdHoCW9AGYQATrAhNgXWfOJldeuIlQYtBewThWeNpSLE1SXlDE3MAFQI9HVCzCysbOQ5Ixx1nZDQsXA9SL0tgXyh-KHoIADczQSrQSAsXNAAmLO1PFXF1FjZOXgFlQIkNaSS-PtUBqXdU2PNva1sOVvTXJsnF-MLi-o6Q4ziU6On4uaSUpcbmnIdZjeVSitNq8GhKfjqMiCaARgIAciQfggcADMBG+UAALKCelAAPQwqDhJ5QfRSEHyP4A+SQ9H-QFcfH4oA))

```typescript
type Replace<Old, New, A extends any[]> =
  A extends [] ? [] :
  A extends [Old, ...infer Rest] ? [New, ...Replace<Old, New, Rest>] :
  A extends [infer T, ...infer Rest] ? [T, ...Replace<Old, New, Rest>] :
  never


type Replace2<Old, New, A extends any[]> =  // Type alias 'Replace' circularly references itself.
  A extends [] ? [] :
  A extends [Old, ...infer Rest] ? [New, Replace2<Old, New, Rest>] :    // Type 'Replace' is not generic.
  A extends [infer T, ...infer Rest] ? [T, Replace2<Old, New, Rest>] :  // Type 'Replace' is not generic.
  never

type A = Replace2<1, 'a', [3, 1, 4, 1]>  // Type is [3, ['a', [4, ['a', []]]]]
```

weswigham commented on src/compiler/checker.ts:
```diff
@@ -10702,6 +10710,9 @@ namespace ts {
                 }
                 if (t.flags & TypeFlags.IndexedAccess) {
+                    if (isGenericTupleType((<IndexedAccessType>t).objectType)) {
+                        return getIndexTypeOfType((<IndexedAccessType>t).objectType, IndexKind.Number);
```
Likewise here, a stringlike index access (eg, "slice") can fetch array apparent members, which are not a component of the index.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -18176,17 +18404,44 @@ namespace ts {
+            return isTupleType(type) ? getTypeReferenceArity(type) - findLastIndex(type.target.elementFlags, f => !!(f & ElementFlags.Variable)) - 1 : 0;
+        }
+
+        function getElementTypeOfTupleType(type: TupleTypeReference, index: number, endSkipCount = 0, writing = false) {
```
```suggestion
        function getElementTypeOfSliceOfTupleType(type: TupleTypeReference, start: number, end: number, writing = false) {

```
might be an easier to grok signature here. What this does (get the combiend element type for a slice of a tuple) isn't obvious from the current name, and the `endSkipCount` parameter feels somewhat awkward, when framed in the context of slicing (why not provide the last element index instead?).

ahejlsberg replied:
Sure, I can rename it. The `endSkipCount` simply ended up being more convenient as we're generally reasoning about elements towards the middle from the start or from the end.

weswigham commented on tests/baselines/reference/elementAccessChain.3.types:
```diff
@@ -170,7 +170,7 @@ for (obj?.a["b"] of []);
 [...obj?.["a"]] = [];
 >[...obj?.["a"]] = [] : never[]
->[...obj?.["a"]] : never[]
+>[...obj?.["a"]] : any[]
```
There a reason for this change?

ahejlsberg replied:
It's an effect of simplifications I made to `checkArrayLiteral`. Given that `obj` is of type `any`, the type of `[...obj?.["a"]]` certainly should be `any[]`. Not sure how the old (and quite convoluted) logic got to `never[]`.

ahejlsberg commented on src/compiler/types.ts:
```diff
@@ -5099,9 +5099,20 @@ namespace ts {
     }
 
+    export const enum ElementFlags {
+        Required  = 1 << 0,  // T
```
It's similar to `TypeFlags` and others. Makes it cheaper to check for multiple values at once.

weswigham replied:
But the most we use at once is two... Is it really worth giving up exhaustiveness and exclusivity checking to replace so few expressions?

ahejlsberg replied:
Yes, I think this is fine, particularly as the current scheme makes it easy for us to add additional element flags that are orthogonal to the kind of element.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -10514,6 +10519,9 @@ namespace ts {
                 }
             }
+            if (isGenericTupleType(type.objectType)) {
+                return getIndexTypeOfType(type.objectType, IndexKind.Number);
```
Yeah, we should only do this when the index type is number-like.

ahejlsberg replied:
So, currently we error on indexed access types with variadic tuple types for any string-like key because we need to do more reasoning about constrains of variadic tuple types. That means you can't actually observe this issue. I think it is fine to leave this as is and address post-beta.

w0rp commented:
This is out of the scope for this pull request, but I think it's the next most obvious thing to ask about for something to come next. Is there an open issue for supporting some way to compute the intersection of all types that are elements of a variadic type tuple, like `T[0] & T[1] & ...`? One use would be to declare `Object.assign` for any number of arguments.

What's interesting is that while I couldn't find a way to compute the intersection, computing the union of all of the elements was easy. I was able to make this pointless and fun function to return a random element.

```ts
const pickOne = <T extends unknown[]>(...args: [...T]): T[number] => {
    return args[Math.floor(Math.random() * args.length)]
}

console.log(pickOne({y: 3}, 4, ['foo'])
```

DanielRosenwasser commented:
Use `T[number]`, pass that through something to [convert a union type to an intersection type](https://stackoverflow.com/questions/50374908/transform-union-type-to-intersection-type).

TylorS commented:
@w0rp There's some small differences between that union to intersection than being able to use `&` directly that are probably worth noting, but may be fine in some cases. Essentially, `UnionToIntersection<A | B>` does not always equal `A & B` - https://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBDAnmYcCqAmAkgHjQPjgF4BYAKDjgAo05QZgA7AEwGc4BDRxOAfmoDWALnQBKYoQBuEAJbM4IxsEnAo4+kzbUqwuDMYAzVXCziiU2c3H8sCuEpVRy5JCjgAVAIzF02HAFdGAUYIAHdGOAAfOABvOAMICEV-AFsAI2MAX3wXZFR3DB9A4LCIgDJY+MTk9KygA

I've come up with an [alternative](https://www.typescriptlang.org/play/#code/AQFwngDgpsBiA2BDEIoDsA8BBANMAysALzACuaA1mgPYDuaAfMcFsFAB6poAmAzsAG0AlmgBmUAE7AAEgF0AsAChgwAPwFgAMhlKVALhZtO6PoJHip0vOcnAAKguVrBCZFwx28hbdIaP9BEq64NDAAKpoAHJQvCAeeLCI8PAARogAxhTM5FR0jMx2RlymiGhgAv7OAN66KnUCANLAIsAUUGDUovayBnaNskUm-MJitnYOzgAU44M8-KXlA+oR0bEedky9dgCUwL39tSoAvgJopAC2KZKVBonJaZlBiiEwdtQAwtRovAAyQmusDjFfgSKCIbhfeBgYALCpMEgVWamLC1dQ5Gj0WoGSaTAB0+MQBiwuyITAWuyBQ2AkxABhsUk8wHxuJAvDpowZ+BJZLKqMEjLen2+fzWdnwfixwDQUAAbpIni97NQAJJoVASXhQdIgIRfbBIkFgiFoKEwspw5grGJxVwodAeD5fX7-OJYBjuhWQV4ARgKKrVkk12t1mAEVWAomo1AMZ0utiOeHDaQkMYuVykR1kDCUQA) using recursive conditional types which gives you more control over the transformation, by being able to customize then `ToConsList` type to work with any specific types your may be using. I've been using this to [aggregate dependencies](https://github.com/TylorS/typed-prelude/blob/master/packages/effects/source/Effect.ts#L53) needed to perform a set of effects. I hope it helps 😄

w0rp commented:
@TylorS Thank you. That is pure genius. That's a pretty perfect solution.

ahejlsberg commented:
@uhyo Regarding the "absorb" behavior in your example [here](https://github.com/microsoft/TypeScript/pull/39094#issuecomment-646113703), this is how you can constrain the type parameter to only permit inference of tuple types:

```ts
declare function pipe<T extends [] | [unknown, ...unknown[]]>(...args: [...T, (...values: [...T]) => void]): void;
```

millsp commented:
This was passing on my tests not long ago, but seems to have broken:

```ts
export type Pop<L extends List> =
    L extends readonly [...infer LBody, any?]
    ? LBody
    : L

type t0 = Pop<[1, 2, 3?]> // now yields `unknown[]`, used to yield `[1, 2]`
```

Not sure if this is intended, so I am reporting anyway.

4.0.0-dev.20200720

TylorS commented:
Might be the wrong place for it, but I'd LOVE LOVE LOVE the above feature to work. It'd allow for generically modelling higher-kinded types in TypeScript without a giant plethora of overloads and fixes other DX-related issues with the current versions in libraries like `fp-ts`

uhyo commented:
@ahejlsberg Thank you for your response!
However, that would still permit types like `[number, ...number[]]` which leads to a runtime error.
To completely prevent the absorb behavior in that example, we need a way to forbid arrays and also unconstrained-length tuple types, which I believe is impossible for now.

ahejlsberg commented:
@millsp Yes, that's an effect of #39281. You can write it like this to get the prior result:

```ts
export type Pop<L extends List> =
    L extends (readonly [...infer LBody, any] | readonly [...infer LBody, any?])
    ? LBody
    : L;
```

You basically have to explicitly acknowledge that you want inferences from both optional and non-optional elements in the last position. It's subtle (as are many inference corner cases), but we'd otherwise have to introduce lower priority speculative inferences, as discussed in #39281, and I'd prefer to avoid that.

ahejlsberg commented:
> However, that would still permit types like [number, ...number[]] which leads to a runtime error.

@uyho True, the only way to truly avoid this behavior would be for us to introduce "rest elements in the middle" support, but that's a non-trivial task and beyond the scope of 4.0.

jdmichaud commented:
Will the type of [`Promise.all`](https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts#L41) be improved thanks to this?

w0rp commented:
> Will the type of [`Promise.all`](https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts#L41) be improved thanks to this?

That's a very good example of something that can easily be improved by using this. Maybe someone should raise an issue or create a pull request for it.

IllusionMH commented:
It's already discussed in #39788

ahejlsberg commented:
With #41544 we now support rest elements in starting and middle positions of tuple types. This addresses some of the issues that were discussed in this PR, such as https://github.com/microsoft/TypeScript/pull/39094#issuecomment-646113703.
